workflows/differentialabundance.nf

```diff
@@ -588,30 +585,83 @@ workflow DIFFERENTIALABUNDANCE {
+        }
+
+    // we parse the functional analysis outputs channel with a tool-based key
+    ch_functional_with_key = ch_functional_results
+        .map {[
+            [method_differential: it[0].method_differential, args_differential: it[0].args_differential, method_functional: it[0].method_functional, args_functional: it[0].args_functional],
+            it.tail()
+        ]}
+        .groupTuple()   // group all the files produced for different contrasts together
+        .cross (
+            ch_tools.map { tools_norm, tools_diff, tools_func -> [
+                // note that the normalized matrix is used to run gsea, whereas differential results are used to run gprofiler2
+                [method_differential: (tools_func.method == 'gsea' ? tools_norm.method : tools_diff.method), args_differential: (tools_func.method == 'gsea' ? tools_norm.args : tools_diff.args), method_functional: tools_func.method, args_functional: tools_func.args],
+                [method_differential: tools_diff.method, args_differential: tools_diff.args, method_functional: tools_func.method, args_functional: tools_func.args]
+            ]}
+        )
+        .map { files, tools ->
+            [tools[1], files.tail()]    // key, [ files ]
+        }
+
+    // we combine the channels defined above with the contrasts, collated versions,
+    // logo, css and citations files. We also combine with the differential outputs
+    // using the tool-based key
+    ch_report_input_files = ch_matrices_with_key
         .combine(VALIDATOR.out.contrasts.map{it.tail()})
         .combine(ch_collated_versions)
         .combine(ch_logo_file)
         .combine(ch_css_file)
         .combine(ch_citations_file)
-        .combine(ch_differential_results.map{it[1]}.toList())
-        .combine(ch_differential_model.map{it[1]}.toList())
-
-    if (params.gsea_run){
-        ch_report_input_files = ch_report_input_files
-            .combine(ch_gsea_results
-                .map{it.tail()}.flatMap().toList()
-            )
-    }
-
-    if (params.gprofiler2_run){
-        ch_report_input_files = ch_report_input_files
-            .combine(gprofiler2_plot_html.map{it[1]}.flatMap().toList())
-            .combine(gprofiler2_all_enrich.map{it[1]}.flatMap().toList())
-            .combine(gprofiler2_sub_enrich.map{it[1]}.flatMap().toList())
-    }
+        .join(ch_differential_with_key)
+        .join(ch_functional_with_key, remainder: true)
+        .map {
+            // functional analysis is optional. When it is not run, the joining with
+            // remainder: true will create a meta with null values for method_functional
+            // and args_functional, and null files. Here we update the meta to keep
+            // args_functional as a map to stay consistent. We also remove null values
+            // from files list.
+            def meta = (it[0].method_functional) ? it[0] : it[0] + [args_functional: [:]]
+            [it[0], it.tail().grep().flatten()]
```

Comment:
could you add a comment describing this channel like you did with the ones before?